File: python/206_reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        # Use extra space with the same size of the linked list
        tmph = ListNode(0)
        h = head
        while h != None:
            new = ListNode(h.val)
            new.next = tmph.next
            tmph.next = new
            h = h.next
        return tmph.next


    def reverseList(self, head):
        # use constant extra space
        newhead = None
        while head != None:
            tmp = head.next
            head.next = newhead
            newhead = head
            head = tmp
        return newhead

File: java/206_reverse_linked_list.java
// https://leetcode.com/problems/reverse-linked-list/

// Reverse a singly linked list.
// A linked list can be reversed either iteratively or recursively. Could you implement both?
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// Iteratively.
public class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null){
            return null;
        }
        if (head != null && head.next == null){
            return head;
        }
        ListNode result = null;
        while (head != null){
            ListNode tmp = head.next;
            head.next = result;
            result = head;
            head = tmp;
        }
        return result;
    }
}